cuDF error handling document (#7917)
rapidsai/cudf: docs/cudf/source/developer_guide/contributing_guide.md

## Exception handling

In alignment with [maintaining compatibility with pandas](#pandas-compatibility),
any API that cuDF shares with pandas should throw all the same exceptions as the
corresponding pandas API given the same inputs.
However, it is not required to match the corresponding pandas API's exception message.

When writing error messages,
sufficient information should be included to help users locate the source of the error,
such as including the expected argument type versus the actual argument provided.

For parameters that are not yet supported,
raise `NotImplementedError`.
There is no need to mention when the argument will be supported in the future.

### Handling libcudf Exceptions

Currently libcudf raises `cudf::logic_error` and `cudf::cuda_error`.

> By default these error types are mapped to `RuntimeError` in python.

Reviewer: ```suggestion
These error types are mapped to `RuntimeError` in Python.
```

Several APIs use the exception payload `what()` message to determine the exception type raised by libcudf.

Determining error type based on exception payload is brittle since libcudf does not maintain API stability on exception messages.

> This is a compromise due to lack of exception diversity in libcudf.

Reviewer: ```suggestion
This is a compromise due to libcudf only raising a limited number of error types.
```

Only adopt this strategy when necessary.

The projected roadmap is to diversify the exception types raised by libcudf.
Standard C++ natively supports various [exception types](https://en.cppreference.com/w/cpp/error/exception),

> which Cython maps to [these Python exception types](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).

Reviewer: ```suggestion
which Cython maps to [these Python exception types](https://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).
```

## Exception handling

In align with [maintaining compatibility with pandas](#pandas-compatibility),
if an API shares an argument with pandas,

> cuDF should raise the same error type given a specific invalid input for that argument.

Author: Agreed, this is more generic!

However, it is not required to match the corresponding pandas API's exception message.

When writing error messages,
sufficient information should be included to help user locate the source of the error,
such as including the expected argument type versus the actual argument provided.

For parameters that are not yet supported,
raise `NotImplementedError`.
There is no need to mention when the argument will be supported in the future.

### Handling libcudf Exceptions

Currently libcudf raises `cudf::logic_error` and `cudf::cuda_error`.
By default these error types are mapped to `RuntimeError` in python.
Several APIs uses the exception payload `what()` message to determine the exception type raised by libcudf.

Determining error type based on exception payload is brittle since libcudf does not maintain API stability on exception messages.
This is a compromise due to lack of exception diversity in libcudf.
Only adopt this strategy when necessary.

The projected roadmap is to diversify the exception types raised by libcudf.
Standard c++ natively supports various [exception types](https://en.cppreference.com/w/cpp/error/exception),

> which follow [these mappings](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions) to python error types.

Reviewer 2: ```suggestion
which Cython maps to [these Python exception types](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).
```

> Custom c++ exceptions types may also be created if necessary.
> Experiments are in progress to create custom cython error mappings to custom exception types in c++.
> This section may be updated as progresses are made on these fronts.

Reviewer 2: ```suggestion
In the future, libcudf may employ custom C++ exception types.
If that occurs, this section will be updated to reflect how these may be mapped to desired Python exception types.
```
